README

Netflix type recommender system competition project on Flixer data from 2005

This is basically an experimentation on how to improve recommendation predictions for movies and users. 

I tried a number of more basic techniques

I incrementally improve on the prediction over time and at this moment have a hybrid user-item based collaborative filtering
system.  The primary similarity algorightm is a mixture of Cosine and Pearson and the prediction algorightm is Resnick we a small
enhancement on user based to include standard deviation. 

This might be useful for others exploring collborative filtering.



changelog.txt
Derek Organ 

CHANGE LOG -

To see all commits and changes you can also view my github repository for this project.
https://github.com/derekorgan/cf 

Note: This github repository is private until week 5 deadline has passed. 

------------------------------------------------------------------------------------
-- WEEK 5
-------------------------------------------------------------------------------------

v0.18 - Z-score normalization
----------------------------------------
https://github.com/derekorgan/cf/commit/14c74b2d30fbe4e4b57d2b1d02973dab3f9ca7ae

While mean-centering removes the offsets caused by the different perceptions of an average 
rating, Z- score normalization also considers the spread in the individual rating scales.

I added to the profile a new method to get the standard deviation. 

I then updated the Resnick function in the user-user based prediction so the standard deviation
is used in the calculation. 

Note: As an reminder you can review my code changes on the github interface.  If you click on 
commits link you can click on a specific commit to view the code changes made. 

RMSE: 0.980464404868937
coverage: 99.87115773973863%


v0.17 - Hybrid Cosine and Pearson Similarity
----------------------------------------
https://github.com/derekorgan/cf/commit/6719195aff1434ecbcf81703bcd2642fae6eccbe

I created a new class CosinePearsonHybrid that averages the cosine similarity with pearson.

I also tried some weighted averages and a section of  my results are below.

...
RMSE: 0.9862317429779974  user Cosine w: 2  user Pearson w:8 item Cosine w:2  item Pearson w:8
RMSE: 0.9848458321532898  user Cosine w: 2  user Pearson w:8 item Cosine w:3  item Pearson w:7
RMSE: 0.9838114250082263  user Cosine w: 2  user Pearson w:8 item Cosine w:4  item Pearson w:6
RMSE: 0.9832487804280765  user Cosine w: 2  user Pearson w:8 item Cosine w:5  item Pearson w:5
RMSE: 0.9831332928989266  user Cosine w: 2  user Pearson w:8 item Cosine w:6  item Pearson w:4
RMSE: 0.98343696653626    user Cosine w: 2  user Pearson w:8 item Cosine w:7  item Pearson w:3
> RMSE: 0.9829833320917658  user Cosine w: 2  user Pearson w:8 item Cosine w:8  item Pearson w:2
RMSE: 0.9835743123981594  user Cosine w: 2  user Pearson w:8 item Cosine w:9  item Pearson w:1
RMSE: 0.9885548189401521  user Cosine w: 3  user Pearson w:7 item Cosine w:1  item Pearson w:9
RMSE: 0.9859485398008404  user Cosine w: 3  user Pearson w:7 item Cosine w:2  item Pearson w:8
RMSE: 0.9846814210724572  user Cosine w: 3  user Pearson w:7 item Cosine w:3  item Pearson w:7
...

The best combination of weighted averages was  

User Based CF -  Cosine Weight :  2     Pearson Weight:  8
Item Base CF -   Cosine Weight :  8     Pearson Weight:  2

RMSE: 0.9829833320917658


v0.16 - Cosine and Pearson combinations for item and user based CF
----------------------------------------

It seems Pearsons with user and Cosine for item based appears to deliver better results.

Item Based         User Based          RMSE
Cosine             Cosine              0.9848579095001863
Pearson 		   Pearson 			   0.9921154630905026
Cosine			   Pearson			   0.9832863160733764
Pearson			   Cosine			   0.9917065477171547	

RMSE: 0.9832863160733767
coverage: 99.77912755383765%

v0.15 - Significance Weighting with Pearson Similarity 
----------------------------------------
https://github.com/derekorgan/cf/commit/eb438efa16753503fa70db4f2d448ab19631c387

Tested Pearson to see if significance weighting would improve it.  After an extensive test of many variations 
I found an improvement on pearson with no significance weighting but not better than the current Cosine numbers
RMSE: 1.0060005117816373     User Significance:1    Item Significance: 1  < no significance weighting
...
RMSE: 0.9921258089791603     User Significance:90    Item Significance: 60
RMSE: 0.9921178026898955     User Significance:90    Item Significance: 65
RMSE: 0.9921154630905042     User Significance:90    Item Significance: 70
RMSE: 0.9921154630905042     User Significance:90    Item Significance: 75
> RMSE: 0.9921154630905026     User Significance:90    Item Significance: 80
RMSE: 0.9921154630905039     User Significance:90    Item Significance: 85
RMSE: 0.9921154630905052     User Significance:90    Item Significance: 90
RMSE: 0.992115463090504     User Significance:90    Item Significance: 95
RMSE: 0.9921154630905036     User Significance:90    Item Significance: 100
...

The best score was with a MAX of 90 for user based and a MAX of 80 for Item.

Pearson without SW      1.0060005117816373
Pearson with SW			0.9921154630905026
Cosine 					0.9848579095001863			

So over all Cosine is still better.

v0.14 - Updated the reader to use the probe data as training data
----------------------------------------

This should help with the predictions on the r.test file because it has more training data to learn from.



------------------------------------------------------------------------------------
-- WEEK 4
-------------------------------------------------------------------------------------

v0.13 - Added significance weighting 
----------------------------------------
https://github.com/derekorgan/cf/commit/7294ca5257671419ec9b4531a85b56cb6e992184

I added significance weighting to the similarity metric for Cosine 

I created a double loop in the ExecuteHybrid to try the ranges of 1-5 and 5-75 in increments of 5 to 
test the best combination of max values for the significance. 

The best combination was no significance for user e.g. MAX = 1 and a max value of 3 for Item 

RMSE: 0.9850593918879242    max user sig :1       max item sig :1
RMSE: 0.9851617636433982    max user sig :1       max item sig :2
RMSE: 0.9848579095001863    max user sig :1       max item sig :3
RMSE: 0.9853052675535504    max user sig :1       max item sig :4
RMSE: 0.9852341878181543    max user sig :2       max item sig :1
RMSE: 0.985335793502781     max user sig :2       max item sig :2
...


src/alg/hybrid/ExecuteHybrid.java
/**
 * Used to evaluate the user-based and item-based hybrid CF algorithm
 * 
 * Derek Organ
 * 27/02/2012
 */

package alg.hybrid;

import java.io.File;

import alg.ib.neighbourhood.*;
import alg.ib.predictor.*;
import alg.ub.neighbourhood.*;
import alg.ub.predictor.*;
import similarity.metric.*;
import util.evaluator.Evaluator;
import util.reader.DatasetReader;

public class ExecuteHybrid
{
	public static void main(String[] args)
	{

		//double bestRMSE = 1000;
		//int n = 10;
		// set the paths and filenames of the item file, train file and test file ...
		String itemFile = "FRT dataset" + File.separator + "r.item";
		String trainFile = "FRT dataset" + File.separator + "r.train";
		String testFile = "FRT dataset" + File.separator + "r.test";
		DatasetReader reader = new DatasetReader(itemFile, trainFile, testFile);
		
		//for (int i=1; i<n; i++){
			//for (int j=1; j<n; j++){
						
		// configure the user-based CF algorithm - set the predictor, neighbourhood and similarity metric ...
		Predictor userPredictor = new Resnick();
		Neighbourhood userNeighbourhood = new NearestNeighbourhood(25);
		SimilarityMetric userMetric = new CosinePearsonHybrid(90, 2, 8);
		//SimilarityMetric userMetric = new Cosine(1);
		
		// configure the item-based CF alogrithm - set the predictor, neighbourhood and similarity metric ...
		PredictorItem itemPredictor = new ResnickItem();
		NeighbourhoodItem itemNeighbourhood = new NearestNeighbourhoodItem(14);
		//SimilarityMetric itemMetric = new Pearson(80); // 80 is the max value for significance weighting	
		SimilarityMetric itemMetric = new CosinePearsonHybrid(80, 8, 2);
		
		//Hybrid Weights for average of user and item based algorithm
		int userWeight = 4;
		int itemWeight = 6;
		
		// set the path and filename of the output file ...
		String outputFile = "results" + File.separator + "predictions.txt";
		
		////////////////////////////////////////////////
		// Evaluates the CF algorithm (do not change!!):
		// - the RMSE (if actual ratings are available) and coverage are output to screen
		// - the output file is created
		
		HybridCF hybrid = new HybridCF(userPredictor, itemPredictor, userNeighbourhood, itemNeighbourhood, userMetric, itemMetric, reader, userWeight, itemWeight);
		Evaluator eval = new Evaluator(hybrid, reader.getTestData());
		
		eval.writeResults(outputFile);
		Double RMSE = eval.getRMSE();
		if(RMSE != null) System.out.println("RMSE: " + RMSE);
		double coverage = eval.getCoverage();
		System.out.println("coverage: " + coverage + "%");
		
		//if(RMSE < bestRMSE) bestRMSE = RMSE;
			//}
		//}
		
		//System.out.println("BestRMSE: " + bestRMSE);

	}
}
	